text/3487-feature-visibility.md

```diff
@@ -0,0 +1,160 @@
+
+# Motivation
+
+[motivation]: #motivation
+
+WIP
+
+No way to hide unstable API such as testing-only features or internally enabled
+features.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+There will be a new flag allowed within `[features]`: `public`. This is boolean
+flag defaulting to `true` that indicates whether or not downstream crates should
+be allowed to use this feature.
+
+```toml
+[features]
+foo = { enables = [], public = false}
+```
+
+Attempting to use a private feature on a downstream crate will result in
+messages like the following:
+
+```
+error: feature `baz` on crate `mycrate` is private and cannot be used by
+  downstream crates
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+`public` is a boolean value that defaults to `true`. It can be thought of as
+`pub` in Rust source files, with the exception of being true by default. If set
+to `false`, Cargo should forbid its use with an error message on any downstream
+crates.
+
+The default `true` is not consistent with [`public_private_dependencies`] or
+Rust's `pub`, but is a reasonable default to be consistent with the current
+behavior. This means that either `feature = []` or
+`feature = { "enables" = [] }` will result in the same configuration.
+
+The name `public` was chosen in favor of `pub` to be consistent with the
+[`public_private_dependencies`] RFC, and to match the existing style of using
+non-truncated words as keys.
+
+In general, marking a feature `public = false` should make tooling treat the
+feature as non-public API. This is described as the following:
+
+- The feature is always usable within the same crate:
+  - Enabled by other features, e.g.
+    `foo = { enables = [some-private-feature] }`, is allowed
+  - Referenced in `[[bench]]` and `[[test]]` target `required-features`
+  - Using the feature on the command-line is allowed
+- Users may explicitly specifying the private features for their dependencies
+  on the command-line (e.g. `--features somecrate/private-feature`) which would
+  otherwise be forbidden
+- The feature should not be accepted by `cargo add --features`
+- The feature should not be reported from `cargo add`'s feature output report
+  - A future tool like `cargo info` shouldn't display information about these
+    features
+- Once `rustdoc` is able to consume feature metadata, `rustdoc` should not
+  document these features unless `--document-private-items` is specified
+
+Attempting to use a private feature in any of the forbidden cases should result
+in an error. Exact details of how features work will likely be refined during
+implementation and experimentation.
+
+Two sample use cases for `public = false` include:
+
+- `docs.rs` having a way to know which features should be hidden
+- Features that are included in feature chains (feature `a` enables feature `b`)
+  but not meant for public consumption could be marked not public
```

Comment:
Seems like this is better served by being in the Motivation

Reply:
Agreed, I reshuffled this information